Master replaced v1 with v2

mirzak opened the pull request:
This is a merge of the rewrite of the `mender-convert` tool. 

See https://tracker.mender.io/browse/MEN-2608 and https://hub.mender.io/t/new-iteration-of-the-mender-convert-tool/824 for additional background information.

The v2 iteration of the tool has the equivalent functionality of v1 (and more).

The following platforms have been run-time tested:

- Raspbian on Raspberry Pi3 3 - U-boot
- Ubuntu Server on Raspberry Pi 3 - U-boot
- Ubuntu Bionic on x86-64 QEMU - GRUB
- Armbian on Pine64 ROCKPro64 (SD and eMMC) - U-boot
     - CONFIG_EFI_STUB not set in Linux kernel which is required for GRUB/EFI integration
- Debian on BBB (SD) - U-boot

Lets quickly go trough the workflow using `mender-convert` v2.

The simplest command of `mender-convert` V2 is the following:

```
MENDER_ARTIFACT_NAME=release-1 ./mender-convert \
    --disk-image input/bone-debian-9.5-iot-armhf-2018-10-07-4gb.img
```

This command will utilize the U-boot - > GRUB integration, and this command should always succeed converting the input image. The output has to be run-time tested to determinate if the U-boot -> GRUB integration will work.

When U-boot -> GRUB integration does not work, one has to fall-back to a U-boot integration. And then the command to convert the image is the following (using BBB as example):

```
MENDER_ARTIFACT_NAME=release-1 ./mender-convert \
     --disk-image input/bone-debian-9.5-iot-armhf-2018-10-07-4gb.img \
     --config configs/beaglebone_black_debian_sdcard_config
```

The `beaglebone_black_debian_sdcard_config` contains all the platform specific code to do an U-boot integration on to the Debian image. Inspect [beaglebone_black_base_config](https://github.com/mirzak/mender-convert/blob/master-replaced-v1-with-v2/configs/beaglebone_black_base_config) and [beaglebone_black_debian_sdcard_config](https://github.com/mirzak/mender-convert/blob/master-replaced-v1-with-v2/configs/beaglebone_black_debian_sdcard_config) for details on how this is done.

Configuration of the tool is done trough configuration files which are supplied using the `--config` option, and multiple `--config` options can be supplied. They will be applied in the order they are provided on the CLI. The [configs/mender_convert_config](https://github.com/mirzak/mender-convert/blob/master-replaced-v1-with-v2/configs/mender_convert_config) is always included and does not need to be supplied as an argument. This file contains all the defaults.

This tool also supports rootfs overlay directories supplied with the `--overlay` option, and multiple overlays can be provided and they will be copied to target rootfs in the order they are provided on the CLI.

The rootfs overlay is used for any configuration that is needed to be done on files in rootfs, e.g if you need want to edit the `/etc/mender/mender.conf` file you will supply your own copy of the file using an rootfs overlay.

The is a convince script provided to bootstrap a `mender.conf` file and a create rootfs directory overlay.

```
./scripts/bootstrap-rootfs-overlay-demo.sh --tenant-token "Paste token here"
```

Above creates,

```
$ tree rootfs_overlay_demo/
rootfs_overlay_demo/
└── etc
    └── mender
        └── mender.conf
```

And this can then be applied using the following command:

```
MENDER_ARTIFACT_NAME=release-1 ./mender-convert \
     --disk-image input/bone-debian-9.5-iot-armhf-2018-10-07-4gb.img \
     --config configs/beaglebone_black_debian_sdcard_config \
     --overlay rootfs_overlay_demo/
```

This will copy the content of `rootfs_overlay_demo` on top of the rootfs during the `mender-convert-modify` stage. So you can also provide files to put on the data partition this way, as rootfs and data are split in the `mender-convert-package` stage.

This PR might not be easy to review, but if you want to take a look at the un-squashed version of v2 you can find it here,

     https://github.com/mirzak/mender-convert/commits/v2

The [README.md](https://github.com/mirzak/mender-convert/blob/master-replaced-v1-with-v2/README.md) contains instructions on how to convert an Raspbian image using a dockerized mender-convert.

Acceptance tests are supported (locally), and you can take a look at the script here,

https://github.com/mirzak/mender-convert/blob/master-replaced-v1-with-v2/scripts/run-tests.sh

Above script is heavily based on [jenkins-mender-convert-build.sh](https://github.com/mendersoftware/mender-qa/blob/master/scripts/jenkins-mender-convert-build.sh)

lluiscampos commented:
I will be looking into this more carefully later today or tomorrow

My first suggestion is to create a `next` branch, rather than master, where to merge this PR until we have the tests running in CI. We cannot merge to master something that is non testable by CI.

This can be a separate task (and I will be happy to take myself).

kacf commented on README.md:
````diff
@@ -0,0 +1,130 @@
+demonstration purposes:
+
+```
+./scripts/bootstrap-rootfs-overlay-demo.sh --tenant-token <paste token from Hosted Mender>
````
Mainly for cognitive reasons, I'm wondering if an `-o` output directory argument should be mandatory for this script. The reason is that it took me some time to understand the relationship between this script and the next command. With a mandatory output directory it would have been immediately obvious.

mirzak replied:
Thanks, updated based on your suggestion

mirzak commented:
> My first suggestion is to create a next branch, rather than master, where to merge this PR until we have the tests running in CI. We cannot merge to master something that is non testable by CI.

This makes sense.

mirzak commented on README.md:
````diff
@@ -0,0 +1,130 @@
+
+```
+sudo apt-get install -y binutils xz-utils file rsync parted e2fsprogs xfsprogs \
+                        pigz dosfstools wget git make bmap-tools
````
Thanks, updated based on your suggestion

mirzak commented on configs/beaglebone_black_base_config:
```diff
@@ -0,0 +1,37 @@
+# Binaries generated with the following script:
+#
+#    https://github.com/drewmoseley/mender-convert-integration-scripts/blob/master/build-uboot-bbb.sh
```
I am not sure we want to do that, but there is a follow-up task for this,

https://tracker.mender.io/browse/MEN-2794

lluiscampos replied:
Good!

mirzak commented on configs/mender_convert_config:
```diff
@@ -0,0 +1,138 @@
+# Mender client version
+#
+# This is used to fetch the correct binaries
+MENDER_CLIENT_VERSION="2.0.1"
```
This relates to https://tracker.mender.io/browse/MEN-2790, and I think we can hold on until we have decided what to do here.

lluiscampos replied:
Good point. Let's handle it then.

mirzak commented on configs/mender_convert_config:
```diff
@@ -0,0 +1,138 @@
+
+source configs/mender_grub_config
+
+mender_create_artifact() {
```
The purpose of it being in `mender_convert_config` is to allow people to override it by providing this function in their own configuration file, e.g to add state-scripts or sign the artifact.

Will clarify this by adding some text

lluiscampos replied:
I didn't see that. It makes sense, then

mirzak commented on scripts/run-tests.sh:
```diff
@@ -0,0 +1,139 @@
+
+    MENDER_ARTIFACT_NAME=${artifact_name} ./mender-convert \
+        --disk-image input/${image_name} \
+        --config /tmp/test_config
```
I will update the test script to run `docker-mender-convert`, and I agree that the `docker-mender-convert` should be the official way of running the tool.

But we really should still support a way of running this tool without docker even though this can be untested (officially), as I know of a case where a Docker container will not work, e.g managing LVM (at least I could not get it work last time I worked with it), where the conversion process had to be done native.

lluiscampos replied:
I agree in keeping the non-docker support as well.

mirzak commented on configs/raspberrypi3_config:
```diff
@@ -0,0 +1,8 @@
+# Binaries generated with the following script:
```
There is an URL to, https://d1b0l86ne08fsf.cloudfront.net/mender-convert/raspberrypi/raspberrypi-integration-scripts.tar.gz

Will not make any changes to this at this point

mirzak commented on configs/rockpro64_emmc_config:
```diff
@@ -0,0 +1,8 @@
+MENDER_STORAGE_DEVICE=/dev/mmcblk1p
+
+# Binaries generated with the following script:
```
There is an URL that points to,

https://d1b0l86ne08fsf.cloudfront.net/mender-convert/armbian/rockpro64/integration-scripts.tar.gz

Will not make any changes to this at this point.

lluiscampos replied:
Sorry, my confusion.

mirzak commented:
@lluiscampos pushed a new "placeholder" commit with updates based on suggestions. The review comments that are still open, I intend not to fix right now and if needed we can create follow-up tasks.

I would suggest that we proceed with this PR. I would like us to squash the commits, I can do this but wanted to give you an opportunity to take a look at the changes I made based on your feedback. With proceeding I mean that we follow your suggestion and put this in a `next` branch for now.

mirzak commented:
> Removing LIC_FILES_CHKSUM.sha256 made the mendertesting/check_license.sh test to fail. Any specific reason you removed it?

I have added this again. No reason for removing it and did not think it would have impact on tests.

mirzak commented on scripts/bootstrap-rootfs-overlay-demo.sh:
```diff
@@ -0,0 +1,64 @@
+  "TenantToken": "${tenant_token}",
+  "UpdatePollIntervalSeconds": 5
+}
+EOF
```
I would like to keep the client configuration separate from the `.deb` package and the primary interface for client configuration in `mender-convert` will be utilizing rootfs overlays. 

I do not see this as a big pain to maintain this (demo purpose), even if there is some duplication. Which to me is ok, if the complexity is lower.

> Any ideas on how to maintain this in the future? Should we split the conf files in the deb package eventually?

I do not see a big win in providing split conf files in the deb package, as the current split is platform specific (the part that is installed in /data), and generally deb packages do not contain platform specific information so we would probably overwrite it anyway.

lluiscampos replied:
Okey. I agree.

Also good point wrt the split of the configuration files.

mirzak commented:
And thank you @lluiscampos for taking time to look in to this and providing valuable feedback

mirzak commented:
I have squashed and changed base. Though I noticed now that the Travis CI failed with,

```
0.25s$ git clone git://github.com/mendersoftware/mendertesting
0.01s$ mendertesting/check_commits.sh
Checking range: next..HEAD:
fatal: ambiguous argument 'next..HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
The command "mendertesting/check_commits.sh" exited with 128.
```

Can I ignore this? @lluiscampos

lluiscampos commented:
@mirzak You can ignore that, yes.

It probably is missing a trick that we use to have in all Travis files. As it has passed before means that the commits are good, so let's merge to next and I'll fix the CI later.
